### ISSUE TITLE
[Proposal] Pass `description` to `ZodError` constructor.

### ISSUE THREAD

**Issue by prescience-data** [colinhacks/zod#935]
After minification / bundling it's not always feasible to access a source map for tracing errors, making it hard to diagnose specific ZodError causes in an app with many Zod schemas. 

It would be handy for the `ZodError` context to include the `description` property (or possibly a new `name` property) of the schema that triggered it by passing this info to the constructor.

### ZodError

https://github.com/colinhacks/zod/blob/12dcdf5169c809e1d3804ed3c2f2fcbf72ee6e54/src/ZodError.ts#L140

```typescript

export class ZodError<T = any> extends Error {
  // ...
  constructor(description: string, issues: ZodIssue[]) {  // Add description to constructor here...
    super();
    // ...
    this.name = "ZodError";
    this.description = description;
    this.issues = issues;
  }
}

```
### handleResult()
https://github.com/colinhacks/zod/blob/12dcdf5169c809e1d3804ed3c2f2fcbf72ee6e54/src/types.ts#L89

```typescript
  if (isValid(result)) {
    return { success: true, data: result.value };
  } else {
    if (!ctx.issues.length) {
      throw new Error("Validation failed but no issues detected.");
    }
    const error = new ZodError(ctx.description, ctx.issues); // Add description to constructor here...
    return { success: false, error };
  }
```

If the schema provided it's `description` to the `ZodError` constructor, it would make it a lot easier to trace issues in these cases where a source map is not available, or even more gracefully in logging with something like:

```typescript
import { EOL } from "node:os"
const logZodError = ({name, description, issues}: ZodError) => {
  const message: string = [
    `Encountered a ${name} error while parsing ${description}`,
     issues.map(
          ({ message, path }: ZodIssue): string =>
            `[${path.join(` › `)}]:  ${message}`
        ).join(EOL)
  ].join(EOL)
  console.error(message)

}

```

**Comment by colinhacks** [comment]
Good idea. Could pair nicely with [this](https://github.com/colinhacks/zod/pull/932).

**Comment by abrouaux** [comment]
I second this proposition, but maybe it'd be better to add the description at the issue level?
Getting back the name of the structure that generated an issue in the logs would be extremely helpful to understand what's going on.

As an example, I'm parsing data in a discriminated union of complex objects with multiple sub levels. 
The backend returning the data is still in flux, so I'm often getting errors in one of the sub level parser.

Previously, using union, I was getting the real error somewhere drowned in all the unionErrors.
With the latest version, I'm getting just the complex object parsing error - but I don't know which one.
